subsys/serialization/transport/rp_ll.c

```diff
@@ -0,0 +1,335 @@
+
+/* Stack size of stack area used by rx thread */
+#define STACKSIZE 1024
+
+/* Scheduling priority used by rx thread */
+#define PRIORITY 7
+
+/* Indicates that handshake was done by this endpoint */
+#define EP_FLAG_HANSHAKE_DONE 1
+
+/* Shared memory configuration */
+#define SHM_START_ADDR      (DT_IPC_SHM_BASE_ADDRESS + 0x400)
+#define SHM_SIZE            0x7c00
+#define SHM_DEVICE_NAME     "sram0.shm"
+
+#define VRING_COUNT         2
+#define VRING_TX_ADDRESS    (SHM_START_ADDR + SHM_SIZE - 0x400)
+#define VRING_RX_ADDRESS    (VRING_TX_ADDRESS - 0x400)
+#define VRING_ALIGNMENT     4
+#define VRING_SIZE          16
+
+#define VDEV_STATUS_ADDR    DT_IPC_SHM_BASE_ADDRESS
+
+
+/* Handlers for TX and RX channels */
+/* TX handler */
+static struct device *ipm_tx_handle;
+
+/* RX handler */
+static struct device *ipm_rx_handle;
+
+static struct rpmsg_virtio_device rvdev;
+struct rpmsg_device *rdev;
+
+static metal_phys_addr_t shm_physmap[] = { SHM_START_ADDR };
+static struct metal_device shm_device = {
+	.name = SHM_DEVICE_NAME,
+	.bus = NULL,
+	.num_regions = 1,
+	.regions = {
+		{
+			.virt = (void *)SHM_START_ADDR,
+			.physmap = shm_physmap,
+			.size = SHM_SIZE,
+			.page_shift = 0xffffffff,
+			.page_mask = 0xffffffff,
+			.mem_flags = 0,
+			.ops = { NULL },
+		},
+	},
+	.node = { NULL },
+	.irq_num = 0,
+	.irq_info = NULL
+};
+
+static struct virtqueue *vq[2];
+
+/* Thread properties */
+static K_THREAD_STACK_DEFINE(rx_thread_stack, STACKSIZE);
+static struct k_thread rx_thread_data;
+
+
+static unsigned char virtio_get_status(struct virtio_device *vdev)
+{
+	return IS_ENABLED(CONFIG_RPMSG_MASTER) ?
+		VIRTIO_CONFIG_STATUS_DRIVER_OK : sys_read8(VDEV_STATUS_ADDR);
+}
+
+static u32_t virtio_get_features(struct virtio_device *vdev)
+{
+	return 0;
+}
+
+void virtio_set_status(struct virtio_device *vdev, unsigned char status)
+{
+	sys_write8(status, VDEV_STATUS_ADDR);
+}
+
+static void virtio_set_features(struct virtio_device *vdev, u32_t features)
+{
+	/* No need for implementation */
+}
+
+static void virtio_notify(struct virtqueue *vq)
+{
+	int status;
+
+	status = ipm_send(ipm_tx_handle, 0, 0, NULL, 0);
+	if (status != 0) {
+		LOG_ERR("Failed to notify: %d", status);
+	}
+}
+
+const struct virtio_dispatch dispatch = {
+	.get_status = virtio_get_status,
+	.get_features = virtio_get_features,
+	.set_status = virtio_set_status,
+	.set_features = virtio_set_features,
+	.notify = virtio_notify,
+};
+
+/* Callback launch right after some data has arrived. */
+static void ipm_callback(void *context, u32_t id, volatile void *data)
+{
+	k_sem_give(&ipm_event_sem);
+}
+
+/* Callback launch right after virtqueue_notification from rx_thread. */
+static int endpoint_cb(struct rpmsg_endpoint *ept, void *data, size_t len,
+		       u32_t src, void *priv)
+{
+	struct rp_ll_endpoint *my_ep = metal_container_of(ept,
+		struct rp_ll_endpoint, rpmsg_ep);
+
+	if (len == 0) {
+		if (!(my_ep->flags & EP_FLAG_HANSHAKE_DONE)) {
```

Comment:
Handshake is very important here. I implemented it, because it is needed. It is not the same as `rpmsg_virtio_wait_remote_ready()`. Please remember that there can be multiple endpoints and handshake is executed on each one at the beginning to make sure that both ends of the endpoint are ready to communicate. `rp_ll_endpoint_init()` is synchronous and it will exit when both sides called `rp_ll_endpoint_init()`, so handshake was done and endpoint is ready to be used.

Reply:
Ah, I see, I missed that the semaphore in the endpoint blocked on init, sorry. I think you might want to check the flag after finishing the k_sem_wait though, that semaphore is yielded for any event.

Related to this, I think the response sending on line 139 should be removed, it's a duplicate of the message sent in endpoint_init. The sample works without it, and even sends one less message across.

Reply:
Message in 139 is also significant. Based on startup timings of cores the first message from net to app may be lost. In this example we get lucky and no messages were lost, but this may not always be the case. Sending two packets from each side ensures that handshake is always correctly done.

Reply:
There is no need for additional check the flag after finishing the k_sem_wait, because at least one of two reasons:
1. receiving any packet (not only handshake packet) is enough to say that endpoint is ready to be used (the other side is ready),
1. There will be not other event before RP_LL_EVENT_CONNECTED